onnxscript/function_libs/torch_lib/ops/core.py

```diff
@@ -3507,10 +3507,41 @@ def aten_linear_backward(
     raise NotImplementedError()
 
 
-def aten_linspace(start: float, end: float, steps: int) -> TensorType:
+@torch_op("aten::linspace", trace_only=True)
+def aten_linspace(
+    start: TRealUnlessFloat16OrInt8, end: TRealUnlessFloat16OrInt8, steps: int, dtype: int = -1
```

Comment:
`start`, `end` and return type can be 3 distinct `TypeVar`s having the same constraints as the underlying types of `TRealUnlessFloat16OrInt8`. @justinchuby what is the current guideline?

Reply:
Do you mean they can take different types?

Reply:
With help of a prototype type constraint deducer, I observe the following. Note I created 3 variants to fold away the if condition, since deducer does not work with trace only functions

```python
"""
NOTE: I'm not sure about this one, the deducer takes INT64 as type for `steps`.
T1: {'tensor(int8)', 'tensor(double)', 'tensor(float)', 'tensor(float16)', 'tensor(uint16)', 'tensor(uint64)', 'tensor(int16)', 'tensor(string)', 'tensor(uint32)', 'tensor(bfloat16)', 'tensor(bool)', 'tensor(uint8)', 'tensor(int32)', 'tensor(int64)'}
T2: {'tensor(int64)'}
start: T1
end: T2
return_val: T2
"""
@torch_op("aten::linspace")
def aten_linspace_variance_1(
    start: TensorType, end: TensorType, steps: int
) -> TensorType:
    zero = op.CastLike(0.0, steps)
    one = op.CastLike(1.0, steps)

    range_tensor = op.Range(zero, steps, one)

    start = op.CastLike(start, end)
    step = op.Div(
        op.Sub(end, start),
        op.Sub(steps, one),
    )

    return op.Add(op.Mul(range_tensor, step), start)


"""
T1: {'tensor(uint16)', 'tensor(int16)', 'tensor(float16)', 'tensor(float)', 'tensor(uint8)', 'tensor(bool)', 'tensor(bfloat16)', 'tensor(string)', 'tensor(uint32)', 'tensor(int64)', 'tensor(int32)', 'tensor(uint64)', 'tensor(int8)', 'tensor(double)'}
T2: {'tensor(uint16)', 'tensor(int16)', 'tensor(float16)', 'tensor(float)', 'tensor(uint8)', 'tensor(bool)', 'tensor(bfloat16)', 'tensor(string)', 'tensor(uint32)', 'tensor(int64)', 'tensor(int32)', 'tensor(uint64)', 'tensor(int8)', 'tensor(double)'}
T3: {'tensor(int16)', 'tensor(float)', 'tensor(int64)', 'tensor(int32)', 'tensor(double)'}
start: T1
end: T2
return_val: T3
"""
@torch_op("aten::linspace")
def aten_linspace_variance_2(
    start: TensorType, end: TensorType, steps: int, dtype: int
) -> TensorType:
    zero = op.Cast(0, to=dtype)
    one = op.Cast(1, to=dtype)
    start = op.Cast(start, to=dtype)
    end = op.Cast(end, to=dtype)
    steps = op.Cast(steps, to=dtype)

    range_tensor = op.Range(zero, steps, one)

    start = op.CastLike(start, end)
    step = op.Div(
        op.Sub(end, start),
        op.Sub(steps, one),
    )

    return op.Add(op.Mul(range_tensor, step), start)

from onnxscript import FLOAT


"""
T1: {'tensor(int16)', 'tensor(uint64)', 'tensor(string)', 'tensor(int32)', 'tensor(uint16)', 'tensor(uint32)', 'tensor(double)', 'tensor(bfloat16)', 'tensor(float16)', 'tensor(int8)', 'tensor(uint8)', 'tensor(bool)', 'tensor(int64)', 'tensor(float)'}
T2: {'tensor(int16)', 'tensor(uint64)', 'tensor(string)', 'tensor(int32)', 'tensor(uint16)', 'tensor(uint32)', 'tensor(double)', 'tensor(bfloat16)', 'tensor(float16)', 'tensor(int8)', 'tensor(uint8)', 'tensor(bool)', 'tensor(int64)', 'tensor(float)'}
T3: {'tensor(float)'}
start: T1
end: T2
return_val: T3
"""
@torch_op("aten::linspace")
def aten_linspace_variance_3(
    start: TensorType, end: TensorType, steps: int,
) -> TensorType:
    zero = op.Cast(0.0, to=FLOAT.dtype)
    one = op.Cast(1.0, to=FLOAT.dtype)
    start = op.Cast(start, to=FLOAT.dtype)
    end = op.Cast(end, to=FLOAT.dtype)
    steps = op.Cast(steps, to=FLOAT.dtype)

    range_tensor = op.Range(zero, steps, one)

    start = op.CastLike(start, end)
    step = op.Div(
        op.Sub(end, start),
        op.Sub(steps, one),
    )

    return op.Add(op.Mul(range_tensor, step), start)
```

Reply:
I see. So we _can_ support different input types, in which case we just create three type vars with the same constraint. However, we can also decide that it’s not practical and they just need to be the same typevar as that’s what torch gives us

Reply:
> it’s not practical and they just need to be the same typevar as that’s what torch gives us

I think different typed `start` and `end` is legal in torch. Return type can be completely different with them too due to `dtype`. So we should use 3 type vars to generate accurate opschema. 

Also assuming deducer is correct, the 1st variant restricts `end` and return value to `INT64` only, hence we might want to insert `steps = op.CastLike(steps, end)` to the beginning. 

```python
"""
T1: {'tensor(int32)', 'tensor(uint64)', 'tensor(uint16)', 'tensor(int8)', 'tensor(bfloat16)', 'tensor(string)', 'tensor(int64)', 'tensor(float16)', 'tensor(uint8)', 'tensor(bool)', 'tensor(double)', 'tensor(uint32)', 'tensor(int16)', 'tensor(float)'}
T2: {'tensor(int32)', 'tensor(int64)', 'tensor(double)', 'tensor(int16)', 'tensor(float)'}
start: T1
end: T2
return_val: T2
"""
@torch_op("aten::linspace")
def aten_linspace_variance_1(
    start: TensorType, end: TensorType, steps: int
) -> TensorType:
    steps = op.CastLike(steps, end)
    zero = op.CastLike(0.0, steps)
    one = op.CastLike(1.0, steps)

    range_tensor = op.Range(zero, steps, one)

    start = op.CastLike(start, end)
    step = op.Div(
        op.Sub(end, start),
        op.Sub(steps, one),
    )

    return op.Add(op.Mul(range_tensor, step), start)
```

Yet the behavior doesn't match with [torch doc](https://pytorch.org/docs/stable/generated/torch.linspace.html#torch.linspace) anyways.
`dtype ([torch.dtype](https://pytorch.org/docs/stable/tensor_attributes.html#torch.dtype), optional) – the data type to perform the computation in. Default: if None, uses the global default dtype (see torch.get_default_dtype()) when both start and end are real, and corresponding complex dtype when either is complex.`

So my comment for this is non-blocking.

Reply:
When I debugged tests of this op, I did see some test data where start is FLOAT and end is INT, so they can have different types in real scenario.

If we have a new solution for such type differences, need to go through all of ops using 'Range' op.

Reply:
Created https://github.com/microsoft/onnxscript/issues/841